### PR TITLE
Valgrind warnings

### PR DESCRIPTION
**Pull request by lgromanowski**
Fixes for "Conditional jump or move depends on uninitialised value(s)" and memleaks reported by valgrind.

### CODE REVIEW

File: apps/openmw/mwsound/ffmpeg_decoder.cpp [hunk 1]
```diff
@@ -211,9 +219,22 @@ void FFmpeg_Decoder::close()
...
+          // avio-close() gives segfault, but with av_free valgrind shows memleak
+          // near mFormatCtx->pb
+          //
+          // to be checked!
+          //
+          // avio_close(mFormatCtx->pb);
+          //
+          av_free(mFormatCtx->pb);
+          mFormatCtx->pb = NULL;
+        }
```
Comment (ghost) [inline apps/openmw/mwsound/ffmpeg_decoder.cpp:246]: This is probably an FFMPEG issue, see https://trac.ffmpeg.org/ticket/1357

File: apps/openmw/mwsound/ffmpeg_decoder.cpp [hunk 2]
```diff
@@ -195,6 +200,9 @@ void FFmpeg_Decoder::open(const std::string &fname)
     }
     catch(std::exception &e)
     {
+        avio_close(mFormatCtx->pb);
```
Comment (ghost) [inline apps/openmw/mwsound/ffmpeg_decoder.cpp]: FFMPEG docs say the AVIOContext should be freed with av_free:
http://ffmpeg.org/doxygen/trunk/avio_8h.html#a853f5149136a27ffba3207d8520172a5

**Comment by ghost** [comment]
We need to investigate the ffmpeg issue more, and the resulting fix (if any) also needs to be applied to the video player.

**Comment by lgromanowski** [comment]
Some memleaks left, but I don't know how to fix them (ie. I don't see nothing wrong at first sight in statsextensions.cpp:1128):

```
==27612== 
==27612== HEAP SUMMARY:
==27612==     in use at exit: 8,807,894 bytes in 3,477 blocks
==27612==   total heap usage: 6,530,596 allocs, 6,527,119 frees, 1,596,186,841 bytes allocated
==27612== 
==27612== 14 bytes in 2 blocks are definitely lost in loss record 8 of 248
==27612==    at 0x4C27730: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==27612==    by 0x9F95709: strdup (in /usr/lib/libc-2.18.so)
==27612==    by 0xA328B66: ??? (in /usr/lib/libX11.so.6.3.0)
==27612==    by 0xA329974: _XimSetICValueData (in /usr/lib/libX11.so.6.3.0)
==27612==    by 0xA324F69: _XimLocalCreateIC (in /usr/lib/libX11.so.6.3.0)
==27612==    by 0xA30B544: XCreateIC (in /usr/lib/libX11.so.6.3.0)
==27612==    by 0x8C1BA0A: ??? (in /usr/lib/libSDL2-2.0.so.0.1.0)
==27612==    by 0x8C1BE39: ??? (in /usr/lib/libSDL2-2.0.so.0.1.0)
==27612==    by 0x8C0E95B: SDL_CreateWindow (in /usr/lib/libSDL2-2.0.so.0.1.0)
==27612==    by 0x6C9732: OEngine::Render::OgreRenderer::createWindow(std::string const&, OEngine::Render::WindowSettings const&) (renderer.cpp:102)
==27612==    by 0x929DA8: OMW::Engine::prepareEngine(Settings::Manager&) (engine.cpp:364)
==27612==    by 0x92A9BD: OMW::Engine::go() (engine.cpp:479)
==27612== 
==27612== 16 bytes in 1 blocks are definitely lost in loss record 18 of 248
==27612==    at 0x4C27CC2: operator new(unsigned long) (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==27612==    by 0x82D489: MWScript::Stats::installOpcodes(Interpreter::Interpreter&) (statsextensions.cpp:1128)
==27612==    by 0x83D22F: MWScript::installOpcodes(Interpreter::Interpreter&, bool) (extensions.cpp:32)
==27612==    by 0x81C29C: MWScript::ScriptManager::run(std::string const&, Interpreter::Context&) (scriptmanagerimp.cpp:109)
==27612==    by 0x83D3B3: MWScript::GlobalScripts::run() (globalscripts.cpp:75)
==27612==    by 0x927D25: OMW::Engine::frameRenderingQueued(Ogre::FrameEvent const&) (engine.cpp:92)
==27612==    by 0x92A307: OMW::Engine::prepareEngine(Settings::Manager&) (engine.cpp:445)
==27612==    by 0x92A9BD: OMW::Engine::go() (engine.cpp:479)
==27612==    by 0x6C85AF: main (main.cpp:275)
==27612== 
==27612== 16 bytes in 1 blocks are definitely lost in loss record 19 of 248
==27612==    at 0x4C27CC2: operator new(unsigned long) (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==27612==    by 0x8389F5: MWScript::Ai::installOpcodes(Interpreter::Interpreter&) (aiextensions.cpp:437)
==27612==    by 0x83D23F: MWScript::installOpcodes(Interpreter::Interpreter&, bool) (extensions.cpp:34)
==27612==    by 0x81C29C: MWScript::ScriptManager::run(std::string const&, Interpreter::Context&) (scriptmanagerimp.cpp:109)
==27612==    by 0x83D3B3: MWScript::GlobalScripts::run() (globalscripts.cpp:75)
==27612==    by 0x927D25: OMW::Engine::frameRenderingQueued(Ogre::FrameEvent const&) (engine.cpp:92)
==27612==    by 0x92A307: OMW::Engine::prepareEngine(Settings::Manager&) (engine.cpp:445)
==27612==    by 0x92A9BD: OMW::Engine::go() (engine.cpp:479)
==27612==    by 0x6C85AF: main (main.cpp:275)
==27612== 
==27612== 24 bytes in 1 blocks are definitely lost in loss record 32 of 248
==27612==    at 0x4C27730: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==27612==    by 0xF4D8C9D: XextCreateExtension (in /usr/lib/libXext.so.6.4.0)
==27612==    by 0x14BA152C: ???
==27612==    by 0x14BA17A0: ???
==27612==    by 0x14BA1407: ???
==27612==    by 0x14B7D026: ???
==27612==    by 0x14B791B0: ???
==27612==    by 0x1464799C: ???
==27612==    by 0x146109B1: ???
==27612==    by 0x14610BCB: ???
==27612==    by 0x14612B94: ???
==27612==    by 0x14663FF8: ???
==27612== 
==27612== 99 bytes in 3 blocks are definitely lost in loss record 151 of 248
==27612==    at 0x4C27730: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==27612==    by 0xA2E6100: XGetWindowProperty (in /usr/lib/libX11.so.6.3.0)
==27612==    by 0x8C1425C: ??? (in /usr/lib/libSDL2-2.0.so.0.1.0)
==27612==    by 0x8C14A07: ??? (in /usr/lib/libSDL2-2.0.so.0.1.0)
==27612==    by 0x8BA6AE6: SDL_PumpEvents (in /usr/lib/libSDL2-2.0.so.0.1.0)
==27612==    by 0x9A9433: SFO::InputWrapper::capture(bool) (sdlinputwrapper.cpp:42)
==27612==    by 0x729E79: MWInput::InputManager::update(float, bool) (inputmanagerimp.cpp:270)
==27612==    by 0x7CFB23: MWGui::LoadingScreen::draw() (loadingscreen.cpp:198)
==27612==    by 0x8C9873: MWWorld::ESMStore::load(ESM::ESMReader&, Loading::Listener*) (esmstore.cpp:114)
==27612==    by 0x8D39EE: MWWorld::EsmLoader::load(boost::filesystem::path const&, int&) (esmloader.cpp:28)
==27612==    by 0x8656B0: MWWorld::GameContentLoader::load(boost::filesystem::path const&, int&) (worldimp.cpp:113)
==27612==    by 0x85D7E4: MWWorld::World::loadContentFiles(Files::Collections const&, std::vector<std::string, std::allocator<std::string> > const&, MWWorld::ContentLoader&) (worldimp.cpp:2047)
==27612== 
==27612== 160 bytes in 1 blocks are definitely lost in loss record 164 of 248
==27612==    at 0x4C27CC2: operator new(unsigned long) (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==27612==    by 0x715B56: MWRender::InventoryPreview::onSetup() (characterpreview.cpp:217)
==27612==    by 0x716E68: MWRender::CharacterPreview::setup() (characterpreview.cpp:97)
==27612==    by 0x79FB98: MWGui::InventoryWindow::updatePlayer() (inventorywindow.cpp:92)
==27612==    by 0x92A236: OMW::Engine::prepareEngine(Settings::Manager&) (engine.cpp:423)
==27612==    by 0x92A9BD: OMW::Engine::go() (engine.cpp:479)
==27612==    by 0x6C85AF: main (main.cpp:275)
==27612== 
==27612== 682 bytes in 1 blocks are definitely lost in loss record 198 of 248
==27612==    at 0x4C27730: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==27612==    by 0x14B7D47B: ???
==27612==    by 0x14B7D8FF: ???
==27612==    by 0x146479C5: ???
==27612==    by 0x146109B1: ???
==27612==    by 0x14610BCB: ???
==27612==    by 0x14612B94: ???
==27612==    by 0x14663FF8: ???
==27612==    by 0x14622B93: ???
==27612==    by 0x511B6EE: Ogre::Root::installPlugin(Ogre::Plugin*) (in /usr/lib/libOgreMain.so.1.9.0)
==27612==    by 0x1466B1C5: ???
==27612==    by 0x511B90A: Ogre::Root::loadPlugin(std::string const&) (in /usr/lib/libOgreMain.so.1.9.0)
==27612== 
==27612== 974 (80 direct, 894 indirect) bytes in 1 blocks are definitely lost in loss record 202 of 248
==27612==    at 0x4C27CC2: operator new(unsigned long) (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==27612==    by 0x9498B5: sh::ScriptLoader::_parseNodes(std::basic_ifstream<char, std::char_traits<char> >&, sh::ScriptNode*) (ScriptLoader.cpp:209)
==27612==    by 0x949D13: sh::ScriptLoader::parseScript(std::basic_ifstream<char, std::char_traits<char> >&) (ScriptLoader.cpp:84)
==27612==    by 0x94A1BF: sh::ScriptLoader::loadAllFiles(sh::ScriptLoader*, std::string const&) (ScriptLoader.cpp:21)
==27612==    by 0x92FF4A: sh::Factory::reloadShaders() (Factory.cpp:668)
==27612==    by 0x9317B3: sh::Factory::loadAllFiles() (Factory.cpp:139)
==27612==    by 0x6EEF02: MWRender::RenderingManager::RenderingManager(OEngine::Render::OgreRenderer&, boost::filesystem::path const&, boost::filesystem::path const&, OEngine::Physic::PhysicEngine*, MWWorld::Fallback*) (renderingmanager.cpp:110)
==27612==    by 0x8643B1: MWWorld::World::World(OEngine::Render::OgreRenderer&, Files::Collections const&, std::vector<std::string, std::allocator<std::string> > const&, boost::filesystem::path const&, boost::filesystem::path const&, ToUTF8::Utf8Encoder*, std::map<std::string, std::string, std::less<std::string>, std::allocator<std::pair<std::string const, std::string> > > const&, int) (worldimp.cpp:223)
==27612==    by 0x92A03C: OMW::Engine::prepareEngine(Settings::Manager&) (engine.cpp:385)
==27612==    by 0x92A9BD: OMW::Engine::go() (engine.cpp:479)
==27612==    by 0x6C85AF: main (main.cpp:275)
==27612== 
==27612== 2,464 bytes in 308 blocks are definitely lost in loss record 223 of 248
==27612==    at 0x4C29D00: memalign (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==27612==    by 0x4C29E17: posix_memalign (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==27612==    by 0x787FAEC: av_malloc (in /usr/lib/libavutil.so.52.48.101)
==27612==    by 0x71A939: MWRender::our_get_buffer(MWRender::AVCodecContext*, MWRender::AVFrame*) (videoplayer.cpp:739)
==27612==    by 0x69D96C1: ??? (in /usr/lib/libavcodec.so.55.39.101)
==27612==    by 0x6650027: ??? (in /usr/lib/libavcodec.so.55.39.101)
==27612==    by 0x69DAD88: avcodec_decode_video2 (in /usr/lib/libavcodec.so.55.39.101)
==27612==    by 0x71EE2E: MWRender::VideoState::video_thread_loop(MWRender::VideoState*) (videoplayer.cpp:768)
==27612==    by 0x5BC1CF9: ??? (in /usr/lib/libboost_thread.so.1.55.0)
==27612==    by 0x8E5E0A1: start_thread (in /usr/lib/libpthread-2.18.so)
==27612==    by 0x9FF93DC: clone (in /usr/lib/libc-2.18.so)
==27612== 
==27612== 18,535 (80 direct, 18,455 indirect) bytes in 1 blocks are definitely lost in loss record 242 of 248
==27612==    at 0x4C27CC2: operator new(unsigned long) (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==27612==    by 0x8868CD1: std::_Rb_tree<unsigned int, std::pair<unsigned int const, MyGUI::GlyphInfo>, std::_Select1st<std::pair<unsigned int const, MyGUI::GlyphInfo> >, std::less<unsigned int>, std::allocator<std::pair<unsigned int const, MyGUI::GlyphInfo> > >::_M_insert_unique(std::pair<unsigned int const, MyGUI::GlyphInfo> const&) (in /usr/lib/libMyGUIEngine.so.3.2.0)
==27612==    by 0x8867568: MyGUI::ResourceManualFont::deserialization(MyGUI::xml::Element*, MyGUI::Version) (in /usr/lib/libMyGUIEngine.so.3.2.0)
==27612==    by 0x808428: MWGui::FontLoader::loadFont(std::string const&) (fontloader.cpp:286)
==27612==    by 0x809D3A: MWGui::FontLoader::loadAllFonts() (fontloader.cpp:139)
==27612==    by 0x75DDB7: MWGui::WindowManager::WindowManager(Compiler::Extensions const&, int, OEngine::Render::OgreRenderer*, std::string const&, std::string const&, bool, Translation::Storage&, ToUTF8::FromType) (windowmanagerimp.cpp:134)
==27612==    by 0x929F8E: OMW::Engine::prepareEngine(Settings::Manager&) (engine.cpp:379)
==27612==    by 0x92A9BD: OMW::Engine::go() (engine.cpp:479)
==27612==    by 0x6C85AF: main (main.cpp:275)
==27612== 
==27612== LEAK SUMMARY:
==27612==    definitely lost: 3,635 bytes in 320 blocks
==27612==    indirectly lost: 19,349 bytes in 247 blocks
==27612==      possibly lost: 0 bytes in 0 blocks
==27612==    still reachable: 8,784,910 bytes in 2,910 blocks
==27612==                       of which reachable via heuristic:
==27612==                         stdstring          : 39,169 bytes in 858 blocks
==27612==         suppressed: 0 bytes in 0 blocks
==27612== Reachable blocks (those to which a pointer was found) are not shown.
==27612== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==27612== 
==27612== For counts of detected and suppressed errors, rerun with: -v
==27612== ERROR SUMMARY: 10 errors from 10 contexts (suppressed: 6 from 3)
```

**Comment by ghost** [comment]
I think I found something -  getSkill and coe have the same opcode:

const int opcodeCOE = 0x200008e;

const int opcodeGetSkill = 0x200008e;

Meaning, only one of them will get deleted?

**Comment by ghost** [comment]
We should also add an assertion to Interpreter::installSegmentX to make sure nothing is already registered for that opcode.

**Comment by lgromanowski** [comment]
Same thing is with opcodeSetHello and opcodeSetFight

**Comment by zinnschlag** [comment]
Oops. But please don't add a comment saying these opcodes are the same. Instead move one of the duplicates to a different one (documentation in apps/openmw/mwscript/docs/vmformat.txt also needs to be updated).

**Comment by lgromanowski** [comment]
Ok, I will do it when I will be back home.

File: apps/openmw/mwrender/videoplayer.cpp [hunk 3]
```diff
@@ -950,8 +950,22 @@ void VideoState::init(const std::string& resourceName)
 
     // Open video file
     /// \todo leak here, ffmpeg or valgrind bug ?
```
Comment (ghost) [inline apps/openmw/mwrender/videoplayer.cpp]: Is there still a leak here now? If not the comment can be removed

Comment (lgromanowski) [inline apps/openmw/mwrender/videoplayer.cpp]: No, it isn't (if format_ctx->pb->buffer is freed), I'll correct that comment.

Comment (ghost) [inline apps/openmw/mwrender/videoplayer.cpp]: Ok. There are two more comments like that, ffmpeg_decoder.cpp:234 and videoplayer.cpp:1026

Comment (lgromanowski) [inline apps/openmw/mwrender/videoplayer.cpp]: Yeah, I've just changed them all.

**Pull request merged**